core-bundle/tests/Functional/ServiceArgumentsTest.php

```diff
@@ -0,0 +1,182 @@
+                        break;
+
+                    case 'tagged_iterator':
+                        $this->assertSame('iterable', $typeName, sprintf('Argument %s of %s is not an iterable', $i, $serviceId));
+                        break;
+
+                    case 'tagged_locator':
+                    case 'service_locator':
+                        $this->assertSame(PsrContainerInterface::class, $typeName, sprintf('Argument %s of %s should be %s but found %s.', $i, $serviceId, PsrContainerInterface::class, $typeName));
+                        break;
+
+                    default:
+                        $this->fail(sprintf('Unknown tagged type "%s" for argument %s ($%s) of service %s.', $parameter->getType(), $i, $parameter->getName(), $serviceId));
+                }
+
+                continue;
+            }
+
+            if (!\is_string($argument)) {
+                if (!$type instanceof \ReflectionNamedType) {
+                    // cannot validate argument without known type
+                    continue;
+                }
+
+                $this->assertTrue($type->isBuiltin() ?? false, sprintf('Argument %s of "%s" should be a built-in type, got "%s".', $i, $serviceId, get_debug_type($argument)));
+
+                if ('iterable' === $typeName) {
+                    $this->assertTrue(is_iterable($argument), sprintf('Argument %s of "%s" is not an iterable.', $i, $serviceId));
+
+                    continue;
+                }
+
+                $this->assertSame(get_debug_type($argument), $typeName);
+                continue;
+            }
+
+            if ($type instanceof \ReflectionNamedType && ('@.inner' === $argument || str_ends_with($argument, '.inner'))) {
+                $this->assertTrue(is_a($class, $typeName, true), sprintf('Argument %s of "%s" should be "%s", got "%s".', $i, $serviceId, $typeName, $class));
+
+                continue;
+            }
+
+            if ('@' === ($argument[0] ?? '')) {
+                $optional = '?' === $argument[1];
+                $service = $container->get(substr($argument, $optional ? 2 : 1), ContainerInterface::NULL_ON_INVALID_REFERENCE);
+
+                if (null === $service) {
+                    $this->assertTrue($optional, sprintf('Unknown service "%s" for argument %s of "%s".', $argument, $i, $serviceId));
+                    $this->assertTrue($parameter->allowsNull(), sprintf('Argument %s of "%s" does not allow NULL but the service "%s" was not found.', $i, $serviceId, $argument));
+
+                    continue;
+                }
+
+                if (!$type instanceof \ReflectionNamedType) {
+                    // cannot validate argument without known type
+                    continue;
+                }
+
+                $this->assertInstanceOf($typeName, $service, sprintf('Argument %s of "%s" should be "%s", got "%s".', $i, $serviceId, $typeName, \get_class($service)));
+            }
+
+            // TODO: handle parameters
```

Comment:
Do you want to add that too? Or ist that for another PR?

Reply:
Added in ddfa7e277a53c61033efde3d27069d17eb5ac7f7